dask/order.py

```diff
@@ -140,37 +155,55 @@ def dependents_key(x):
                 if d not in result and not (d in seen and len(waiting[d]) > 1)]
         if len(deps) < 1000:
             deps = sorted(deps, key=dependents_key, reverse=True)
+
         stack.extend(deps)
 
     return result
 
 
-def ndependents(dependencies, dependents):
+def ndependents(dependencies, dependents, total_dependencies):
     """ Number of total data elements that depend on key
 
-    For each key we return the number of data that can only be run after this
+    For each key we return the number of keys that can only be run after this
     key is run.  The root nodes have value 1 while deep child nodes will have
     larger values.
 
+    We also return the minimum value of the maximum number of dependencies of
+    all final dependencies (see module-level comment for more)
+
     Examples
     --------
     >>> dsk = {'a': 1, 'b': (inc, 'a'), 'c': (inc, 'b')}
     >>> dependencies, dependents = get_deps(dsk)
 
-    >>> sorted(ndependents(dependencies, dependents).items())
+    >>> total_dependencies = ndependencies(dependencies, dependents)
+    >>> total_dependents, min_dependencies = ndependents(dependencies,
+    ...                                                  dependents,
+    ...                                                  total_dependencies)
+    >>> sorted(total_dependents.items())
     [('a', 3), ('b', 2), ('c', 1)]
+
+    Returns
+    -------
+    total_dependendents: Dict[key, int]
+    min_dependencies: Dict[key, int]
     """
     result = dict()
+    min_result = dict()
     num_needed = {k: len(v) for k, v in dependents.items()}
     current = {k for k, v in num_needed.items() if v == 0}
     while current:
         key = current.pop()
         result[key] = 1 + sum(result[parent] for parent in dependents[key])
+        try:
```

Comment:
Is the `try / except` to handle the empty case? If so, `min` takes a `default=` argument. So

```python
min_result[key] = min(min_result[parent] for parent in dependents[key], default=total_dependencies[key])
```

Reply:
While more verbose, try-except is often performance optimal when the try case is much more common.  

```python
In [1]: d = {'x': 1}

In [2]: %timeit min([], default=d['x'])
378 ns ± 16.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: %timeit min([1, 2], default=d['x'])
413 ns ± 1.26 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [4]: %timeit min([1, 2])
166 ns ± 0.642 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [6]: %%timeit
   ...: try:
   ...:     min([1, 2])
   ...: except TypeError:
   ...:     d['x']
   ...:     
178 ns ± 2.74 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [8]: %%timeit
   ...: try:
   ...:     min([])
   ...: except ValueError:
   ...:     d['x']
   ...:     
493 ns ± 4.52 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

This difference is small, but we do actually care about things at the microsecond level here.

Reply:
Huh, I'm surprised by the overhead added by the default argument. Good to know.